src/java/org/apache/cassandra/db/commitlog/CommitLog.java

```diff
@@ -234,7 +237,16 @@ public Map<Keyspace, Integer> recoverSegmentsOnDisk(ColumnFamilyStore.FlushReaso
             logger.info("Log replay complete, {} replayed mutations", replayedKeyspaces.values().stream().reduce(Integer::sum).orElse(0));
 
             for (File f : files)
-                segmentManager.handleReplayedSegment(f);
+            {
+                if(!segmentsWithInvalidMutations.get(f.name()).isEmpty())
+                {
+                    logger.debug("File {} should not be deleted as it contains invalid mutations", f.name());
```

Comment:
The "decision" to delete the segment is actually taken by `CommitLogSegmentManagerStandard`: so should we maybe pass a boolean to `handleReplayedSegment` (i.e. `hasInvalidMutations`) and eventually skip the deletion there?